src/gt4py/next/program_processors/runners/dace_iterator/itir_to_tasklet.py

```diff
@@ -184,49 +184,49 @@ def __init__(
     offset_literal, data = node_args
     assert isinstance(offset_literal, itir.OffsetLiteral)
     offset_dim = offset_literal.value
     assert isinstance(offset_dim, str)
+    offset_provider = transformer.offset_provider[offset_dim]
+    assert isinstance(offset_provider, NeighborTableOffsetProvider)
+
     iterator = transformer.visit(data)
     assert isinstance(iterator, IteratorExpr)
     field_desc = iterator.field.desc(transformer.context.body)
-
-    field_index = "__field_idx"
-    offset_provider = transformer.offset_provider[offset_dim]
-    if isinstance(offset_provider, NeighborTableOffsetProvider):
-        neighbor_check = f"{field_index} >= 0"
-    elif isinstance(offset_provider, StridedNeighborOffsetProvider):
-        neighbor_check = f"{field_index} < {field_desc.shape[offset_provider.neighbor_axis.value]}"
-    else:
-        assert isinstance(offset_provider, Dimension)
-        raise NotImplementedError(
-            "Neighbor reductions for cartesian grids not implemented in DaCe backend."
-        )
+    origin_index_node = iterator.indices[offset_provider.origin_axis.value]
 
     assert transformer.context.reduce_identity is not None
+    assert transformer.context.reduce_identity.dtype == iterator.dtype
 
-    sdfg: dace.SDFG = transformer.context.body
-    state: dace.SDFGState = transformer.context.state
-
-    shifted_dim = offset_provider.origin_axis.value
-
-    result_name = unique_var_name()
+    # gather the neighbors in a result array dimensioned for `max_neighbors`
+    neighbor_value_var = unique_var_name()
     sdfg.add_array(
-        result_name, dtype=iterator.dtype, shape=(offset_provider.max_neighbors,), transient=True
+        neighbor_value_var,
+        dtype=iterator.dtype,
+        shape=(offset_provider.max_neighbors,),
+        transient=True,
     )
-    result_access = state.add_access(result_name, debuginfo=di)
+    neighbor_value_node = state.add_access(neighbor_value_var, debuginfo=di)
+
+    # allocate scalar to store index for direct addressing of neighbor field
+    neighbor_index_var = unique_var_name()
+    sdfg.add_scalar(neighbor_index_var, _INDEX_DTYPE, transient=True)
+    neighbor_index_node = state.add_access(neighbor_index_var, debuginfo=di)
 
     # generate unique map index name to avoid conflict with other maps inside same state
-    neighbor_index = unique_name("neighbor_idx")
+    neighbor_map_index = unique_name(f"{offset_dim}_neighbor_map_idx")
     me, mx = state.add_map(
-        f"{offset_dim}_neighbors_map",
-        ndrange={neighbor_index: f"0:{offset_provider.max_neighbors}"},
+        f"{offset_dim}_neighbor_map",
+        ndrange={neighbor_map_index: f"0:{offset_provider.max_neighbors}"},
         debuginfo=di,
     )
     table_name = connectivity_identifier(offset_dim)
-    table_subset = (f"0:{sdfg.arrays[table_name].shape[0]}", neighbor_index)
+    table_subset = (f"0:{sdfg.arrays[table_name].shape[0]}", neighbor_map_index)
```

Comment:
I do not get why this is `.shape[0]` since this are all cells and not the neighbors.

Reply:
We let the memlet cut the slice for `neighbor_map_index` and then `shift_tasklet` will do access based on cell index. This could be an issue for DaCe, I see now. The tasklet should receive the array full shape, so it can treat the array as a C pointer and access it with all explicit indexes.